text/0605-eks-rewrite.md

```diff
@@ -0,0 +1,476 @@
+  outputClusterName: true,
+  outputConfigCommand: true,
+});
+```
+
+After
+
+```
+const cluster = new eks.Cluster(this, 'Cluster', {
+  version: eks.KubernetesVersion.V1_31,
+  outputInfo: true,
+});
+```
+
+### Kubectl Handler Configuration
+
+KubectlProvider is a lambda function that CDK deploys alongside the EKS cluster
+in order to execute kubectl commands against the cluster.
+
+A common scenarios is that users create a CDK app that deploys the EKS cluster,
+which is then imported in other apps in order to deploy resources onto the
+cluster.
+
+Difference
+
+Before
+
+```
+const kubectlProvider = eks.KubectlProvider.fromKubectlProviderAttributes(this, 'KubectlProvider', {
+  functionArn,
+  kubectlRoleArn: 'arn:aws:iam::123456789012:role/kubectl-role',
+  handlerRole,
+});
+```
+
+After
+
+```
+const kubectlProvider = eks.KubectlProvider.fromKubectlProviderArn(this, 'KubectlProvider', {
+  functionArn: // Required. ARN of the original kubectl function
+});
+```
+
+Following parameters are removed:
+
+- kubectlRoleArn
+- handlerRole
+
+`fromKubectlProviderAttributes()` is renamed to
+`fromKubectlProviderFunctionArn()`.
+
+Reason: when the KubectlProvider was created in another stack, the lambda
+execution role already has permissions to access the cluster.
+
+Besides that, optional KubectlProvider specific properties are moved into
+KubectlProviderOptions to better organize properties.
+
+KubectlProviderOptions Definition
+
+```
+export interface KubectlProviderOptions {
+  readonly role?: iam.IRole;
+  readonly awscliLayer?: lambda.ILayerVersion;
+  readonly kubectlLayer?: lambda.ILayerVersion;
+  readonly memory?: Size;
+  readonly environment?: { [key: string]: string };
+  /**
+   * Which subnets should the provider functions be placed in.
+   */
+  readonly vpcSubnets?: ec2.SubnetSelection;
+}
+```
+
+Usage - Before
+
+```
+new eks.Cluster(this, 'MyCluster', {
+  version: eks.KubernetesVersion.V1_31,
+  kubectlMemory: Size.gigabytes(4),
+  kubectlLayer: new KubectlV31Layer(this, 'kubectl'),
+  kubectlEnvironment: {
+    'http_proxy': 'http://proxy.myproxy.com',
+  },
+  kubectlRole: iam.Role.fromRoleArn(
+    this,
+    'MyRole',
+    'arn:aws:iam::123456789012:role/lambda-role'
+  );
+});
+```
+
+Usage - After
+
+```
+new eks.Cluster(this, 'MyCluster', {
+  version: eks.KubernetesVersion.V1_31,
+  kubectlLayer: new KubectlV31Layer(this, 'kubectl'),
+  kubectlProviderOptions: {
+    memory: Size.gigabytes(4),
+    environment: {
+      'http_proxy': 'http://proxy.myproxy.com',
+    },
+    role: iam.Role.fromRoleArn(
+      this,
+      'MyRole',
+      'arn:aws:iam::123456789012:role/lambda-role'
+    );
+});
+```
+
+## Migration Path
+
+Note: We can't guarantee it's a safe migration.
+
+Due to the fact that switching from a custom resource
+(Custom::AWSCDK-EKS-Cluster) to a native L1 (AWS::EKS::Cluster) resource
+requires cluster replacement, CDK users who need to preserve their cluster will
+have to take additional actions.
```

Comment:
We will need a blog post about this I think. Also - lets consider creating a dedicated migration tool for customers to use, this is too many manual operations.

Reply:
Please add explanations to each of these steps.